client/app/rollback.go

```diff
@@ -2,75 +2,63 @@ package app
 
 import (
 	"context"
-	"fmt"
 	"math/big"
 
-	"cosmossdk.io/store"
-	storemetrics "cosmossdk.io/store/metrics"
-
 	cmtcmd "github.com/cometbft/cometbft/cmd/cometbft/commands"
-	dbm "github.com/cosmos/cosmos-db"
+	cmtcfg "github.com/cometbft/cometbft/config"
 	"github.com/ethereum/go-ethereum/common"
 
 	"github.com/piplabs/story/lib/errors"
-	"github.com/piplabs/story/lib/log"
 )
 
-func Rollback(ctx context.Context, cfg Config, removeBlock bool) error {
-	engineCl, err := newEngineClient(ctx, cfg)
-	if err != nil {
-		return err
+func RollbackCometAndAppState(ctx context.Context, a *App, appCfg Config, cometCfg cmtcfg.Config, rollbackHeights int64, removeBlock bool, rollbackEVM bool) (lastHeight int64, lastHash []byte, err error) {
+	for range rollbackHeights {
+		lastHeight, lastHash, err = cmtcmd.RollbackState(&cometCfg, removeBlock)
+		if err != nil {
+			return lastHeight, lastHash, errors.Wrap(err, "failed to rollback CometBFT state")
+		}
 	}
 
-	latestHeigth, err := engineCl.BlockNumber(ctx)
-	if err != nil {
-		return err
+	if err = a.CommitMultiStore().RollbackToVersion(lastHeight); err != nil {
+		return lastHeight, lastHash, errors.Wrap(err, "failed to rollback to version")
 	}
 
-	latestBlock, err := engineCl.BlockByNumber(ctx, big.NewInt(int64(latestHeigth)))
-	if err != nil {
-		return err
-	} else if latestBlock.BeaconRoot() == nil {
-		return errors.New("cannot rollback EVM with nil beacon root", "height", latestHeigth)
+	if !rollbackEVM || rollbackHeights <= 1 {
+		return lastHeight, lastHash, nil
 	}
 
-	db, err := dbm.NewDB("application", cfg.BackendType(), cfg.DataDir())
+	engineCl, err := newEngineClient(ctx, appCfg)
```

Comment:
I think based on what we discussed with Omni and this [PR](https://github.com/omni-network/omni/pull/1072/files), we cannot support EVM rollback and hence this part of the code was removed by Omni

Reply:
reflected in e78cb90